src/couch/src/couch_debug.erl

```diff
@@ -620,6 +676,90 @@ info_size(InfoKV) ->
         {binary, BinInfos} -> lists:sum([S || {_, S, _} <- BinInfos]);
         {_, V} -> V
     end.
+resource_hoggers(MemoryInfo, InfoKey) ->
+    KeyFun = fun
+        ({_Pid, _Id, undefined}) -> undefined;
+        ({_Pid, Id, DataMap}) -> {Id, [{InfoKey, maps:get(InfoKey, DataMap)}]}
+    end,
+    resource_hoggers(MemoryInfo, InfoKey, KeyFun).
+
+resource_hoggers(MemoryInfo, InfoKey, KeyFun) ->
+    HoggersData = resource_hoggers_data(MemoryInfo, InfoKey, KeyFun),
+    TableSpec = [
+        {50, centre, id},
+        {20, centre, InfoKey}
+    ],
+    print_table(HoggersData, TableSpec).
+
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun) when is_atom(InfoKey) ->
+    resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, 20).
+
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, N) when is_atom(InfoKey) and is_integer(N) ->
+    SortedTuples = resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, undefined),
+    {TopN, _} = lists:split(N, SortedTuples),
+    TopN;
+resource_hoggers_data(MemoryInfo, InfoKey, KeyFun, undefined) when is_atom(InfoKey) ->
+    Tuples = lists:filtermap(
+        fun(Tuple) ->
+            case KeyFun(Tuple) of
+                undefined ->
+                    false;
+                Value ->
+                    {true, Value}
+            end
+        end,
+        MemoryInfo
+    ),
+    lists:reverse(lists:keysort(2, Tuples)).
+
+resource_hoggers_snapshot({N, MemoryInfo, InfoKeys} = _Snapshot) ->
+    Data = lists:filtermap(
+        fun({Pid, Id, Data}) ->
+            case memory_info(Pid, InfoKeys) of
+                {Pid, undefined, undefined} ->
+                    false;
+                {_, _, DataMap} ->
+                    {true, {Pid, Id, update_delta(Data, DataMap)}}
+            end
+        end,
+        MemoryInfo
+    ),
+    {N + 1, Data, InfoKeys};
+resource_hoggers_snapshot([]) ->
+    [];
+resource_hoggers_snapshot([{_Pid, _Id, Data} | _] = MemoryInfo) ->
+    resource_hoggers_snapshot({0, MemoryInfo, maps:keys(Data)}).
+
+update_delta({_, InitialDataMap}, DataMap) ->
+    update_delta(InitialDataMap, DataMap);
+update_delta(InitialDataMap, DataMap) ->
+    Delta = maps:fold(
+        fun(Key, Value, AccIn) ->
+            maps:put(Key, maps:get(Key, DataMap, Value) - Value, AccIn)
+        end,
+        maps:new(),
+        InitialDataMap
+    ),
+    {Delta, InitialDataMap}.
+
+analyze_resource_hoggers({N, Data, InfoKeys}, TopN) ->
+    io:format("Number of snapshots: ~p~n", [N]),
+    lists:map(
+        fun(InfoKey) ->
+            KeyFun = fun
+                ({_Pid, _Id, undefined}) -> undefined;
+                ({_Pid, Id, {_Delta, DataMap}}) -> {Id, [{InfoKey, maps:get(InfoKey, DataMap)}]}
```

Comment:
This doesn't use `Delta`. The idea was to get information about the change of value of InfoKey to detect a leak. The constant increase of memory consumption would be manifested by big value of Delta. Then the call to `resource_hoggers_data` would take the top N entries.

Reply:
Ideally we would display an ordered by delta table 

```
| id | initial | current | delta
```

However it might complicate things a lot. Therefore out of scope for this PR. Unless there is magic no more then 5-10 lines which would do it.

Reply:
For example, the first map in the following is the `Delta` and the second map is `DataMap`. This is from a call to `resource_hoggers_snapshot`.
```
{1,
 [{<0.0.0>,"init[<0.0.0>]",
   {#{binary => 0,dictionary => 0,heap_size => 0,links => 0,
      memory => 0,message_queue_len => 0,monitored_by => 0,
      monitors => 0,stack_size => 0,total_heap_size => 0},
    #{binary => 0,dictionary => 0,heap_size => 2586,links => 3,
      memory => 34384,message_queue_len => 0,monitored_by => 0,
      monitors => 0,stack_size => 3,total_heap_size => 4188}}},
  {<0.1.0>,"erts_code_purger[<0.1.0>]",
   {#{binary => 0,dictionary => 0,heap_size => 0,links => 0,
      memory => 0,message_queue_len => 0,monitored_by => 0,
      monitors => 0,stack_size => 0,total_heap_size => 0},
    #{binary => 0,dictionary => 0,heap_size => 233,links => 0,
      memory => 3256,message_queue_len => 0,monitored_by => 0,
      monitors => 0,stack_size => 2,total_heap_size => 312}}},
...
```

Reply:
> That is, it should be {Id, [{InfoKey, maps:get(InfoKey, Delta}]}?

yes